typing/ctype.ml

```diff
@@ -1809,6 +1878,199 @@ let expand_head_opt env ty =
+   in some edge cases (when [get_unboxed_type_representation] ran out of fuel,
+   or when the type is a Tconstr that is missing from the Env). *)
+let rec estimate_type_layout env ty =
+  let ty = repr ty in
+  let open Type_layout in
+  match ty.desc with
+  | Tconstr(p, _, _) -> begin
+      match Env.find_type p env with
+      | { type_kind = k } -> Layout (layout_bound_of_kind k)
+      | exception Not_found -> Layout any
+    end
+  | Tvariant row ->
+      let row = Btype.row_repr row in
+      (* if all labels are devoid of arguments, not a pointer *)
+      if
+        not row.row_closed
+        || List.exists
+          (function
+            | _, (Rpresent (Some _) | Reither (false, _, _, _)) -> true
+            | _ -> false)
+          row.row_fields
+      then Layout value
+      else Layout immediate
+  | Tvar { layout; _ } -> Var (layout, ty)
+  | Tarrow _ -> Layout value
+  | Ttuple _ -> Layout value
+  | Tobject _ -> Layout value
+  | Tfield _ -> Layout value
+  | Tnil -> Layout value
+  | (Tlink _ | Tsubst _) -> assert false
+  | Tunivar { layout; _ } -> Layout layout
+  | Tpoly (ty, _) -> estimate_type_layout env ty
+  | Tpackage _ -> Layout value
+
+(* For convenience, returns the most precise layout we computed for the type
+   (which may still be an upper bound). *)
+let rec constrain_type_layout ~fixed env ty1 layout2 =
+  let constrain_unboxed ty1 =
+    match estimate_type_layout env ty1 with
+    | Layout layout1 -> Type_layout.sublayout layout1 layout2
+    | Var (layout1, ty) ->
+      if fixed then Type_layout.sublayout layout1 layout2
+      else
+        Result.map (fun layout1 -> set_var_layout ty layout1; layout1)
+          (Type_layout.intersection layout1 layout2)
+  in
+  (* This is an optimization to avoid calling [get_unboxed_type_representation]
+     if we can tell the constraint is satisfied from the type_kind *)
+  match (repr ty1).desc with
+  | Tconstr(p, _args, _abbrev) -> begin
+      let layout_bound =
+        begin match Env.find_type p env with
+        | { type_kind = k; _ } -> Type_layout.layout_bound_of_kind k
+        | exception Not_found -> Type_layout.any
+        end
+      in
+      match Type_layout.sublayout layout_bound layout2 with
+      | Ok _ as ok -> ok
+      | Error _ ->
+        (* CR ccasinghino: Can we improve performance by reimplementing
+           get_unboxed_type_representation and adding a layout check at each
+           iteration, rather than fully expanding? *)
+        let ty1 = get_unboxed_type_representation env ty1 in
+        constrain_unboxed ty1
+    end
+  | Tpoly (ty, _) -> constrain_type_layout ~fixed env ty layout2
+  | _ -> constrain_unboxed ty1
+
+let check_type_layout env ty layout =
+  constrain_type_layout ~fixed:true env ty layout
+
+let constrain_type_layout env ty layout =
+  constrain_type_layout ~fixed:false env ty layout
+
+let check_decl_layout env decl layout =
+  match decl with
+  | { type_kind =
```

Comment:
I think not, and I've deleted it.

Slightly longer story:  This case was needed in the immediacy rework, because there we don't have fully accurate information in the kinds for unboxed types.  But here we shouldn't need it, because we store a layout for unboxed types and ensure it's accurate at the end of type declaration checking.  I am slightly worried that it may not be fully accurate in cases involving complicated mutually recursive types, but that's a separate problem that's tracked there.  It would only result in being conservative here, not unsound, and none of our current test cases are tripped up by it, so I think it's fine to delete this and we can add it back if we find an actual example of this hypothetical problem that for some reason can't be resolved when we check the declaration.